src/main/scala/viper/gobra/frontend/Parser.scala

```diff
@@ -65,7 +89,109 @@ object Parser {
+        (res, positions)
+      }
+      val (res, pos) = sourceCache.getOrElse(getCacheKey(source), parseAndStore())
+      if (cacheHit) {
+        // a cached AST has been found in the cache. The position manager does not yet have any positions for nodes in
+        // this AST. Therefore, the following strategy iterates over the entire AST and copies positional information
+        // from the cached positions to the position manager
+        val copyPosStrategy = strategyWithName[Any]("copyPositionInformation", {
+          case n: PNode =>
+            val start = pos.getStart(n)
+            val finish = pos.getFinish(n)
+            start.foreach(positions.setStart(n, _))
+            finish.foreach(positions.setFinish(n, _))
+            Some(n): Option[Any]
+          case n => Some(n)
+        })
+        res.map(prog => rewrite(topdown(copyPosStrategy))(prog))
+      } else {
+        res
+      }
+    }
+
+    def isErrorFree(parserResults: Vector[Either[Vector[ParserError], PProgram]]): Either[Vector[ParserError], Vector[PProgram]] = {
+      val (errors, programs) = parserResults.partitionMap(identity)
+      if (errors.isEmpty) Right(programs) else Left(errors.flatten)
+    }
+
+    def samePackage(programs: Vector[PProgram]): Either[Vector[ParserError], Vector[PProgram]] = {
+      require(programs.nonEmpty)
+      val pkgName = programs.head.packageClause.id.name
+      val differingPkgNameMsgs = programs.flatMap(p =>
+        error(
+          p.packageClause,
+          s"Files have differing package clauses, expected $pkgName but got ${p.packageClause.id.name}",
+          p.packageClause.id.name != pkgName))
+      if (differingPkgNameMsgs.isEmpty) Right(programs) else Left(pom.translate(differingPkgNameMsgs, ParserError))
+    }
+
+    def makePackage(programs: Vector[PProgram]): Either[Vector[ParserError], PPackage] = {
+      val clause = rewriter.deepclone(programs.head.packageClause)
+      val parsedPackage = PPackage(clause, programs, pom)
+      // The package parse tree node gets the position of the package clause:
+      pom.positions.dupPos(clause, parsedPackage)
+      Right(parsedPackage)
+    }
+
+    val parsingFn = if (config.cacheParser) { parseSourceCached _ } else { parseSource _ }
+    val parserResults = sources.map(parsingFn)
+    val res = for {
+      // check that each of the parsed programs has the same package clause. If not, the algorithm collecting all files
+      // of the same package has failed or users have entered an invalid collection of inputs
+      programs <- isErrorFree(parserResults)
+      programs <- samePackage(programs)
+      pkg <- makePackage(programs)
+    } yield pkg
+    // report potential errors:
+    res.left.map(errors => {
+      val groupedErrors = errors.groupBy{ _.position.get.file }
+      groupedErrors.foreach { case (p, pErrors) =>
+        config.reporter report ParserErrorMessage(p, pErrors)
+      }
+      errors
+    })
```

Comment:
I think this should be something along the lines of:
```suggestion
    res.left.foreach(errors => {
      val groupedErrors = errors.groupBy{ _.position.get.file }
      groupedErrors.foreach { case (p, pErrors) =>
        config.reporter report ParserErrorMessage(p, pErrors)
      }
    })
    
    res
```